### PR TITLE
Extract rpc/tensorpipe_utils.{cpp,h} from rpc/utils.{cpp,h}

### PR DESCRIPTION
**Pull request by beauby**
Test Plan: CI

Differential Revision: D23732022

### CODE REVIEW

**Comment by lw** [comment]
CI is failing because of
```
Sep 17 10:28:36 ../lib/libtensorpipe_agent.so: undefined reference to `torch::jit::getWriteableTensorData(at::Tensor const&)'
Sep 17 10:28:36 ../lib/libtensorpipe_agent.so: undefined reference to `torch::jit::Pickler::protocol()'
Sep 17 10:28:36 ../lib/libtensorpipe_agent.so: undefined reference to `torch::jit::Pickler::~Pickler()'
Sep 17 10:28:36 ../lib/libtensorpipe_agent.so: undefined reference to `torch::jit::Pickler::pushIValue(c10::IValue const&)'
Sep 17 10:28:36 ../lib/libtensorpipe_agent.so: undefined reference to `torch::jit::Pickler::stop()'
Sep 17 10:28:36 ../lib/libtensorpipe_agent.so: undefined reference to `torch::jit::Unpickler::parse_ivalue()'
Sep 17 10:28:36 clang: error: linker command failed with exit code 1 (use -v to see invocation)
Sep 17 10:28:36 test_cpp_rpc/CMakeFiles/test_cpp_rpc.dir/build.make:177: recipe for target 'bin/test_cpp_rpc' failed
Sep 17 10:28:36 make[2]: *** [bin/test_cpp_rpc] Error 1
Sep 17 10:28:36 make[1]: *** [test_cpp_rpc/CMakeFiles/test_cpp_rpc.dir/all] Error 2
```

**Review by lw — APPROVED** [review]
Looks good to me. I'd however like for @mrshenli and for @malfet to also check if possible.

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -299,7 +299,12 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     # easier to other targets that depend on this. However, in that case, by not
     # setting the USE_TENSORPIPE compile definition, this target will just end
     # up being empty. Downstream targets should also add a #ifdef guard.
-    add_library(tensorpipe_agent "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.cpp" "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.h")
+    add_library(tensorpipe_agent
+      "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.cpp"
+      "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.h"
+      "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_utils.cpp"
+      "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_utils.h"
```
Comment (malfet) [inline caffe2/CMakeLists.txt:306]: There is no need to add header files to the library, is there?
Also, if you expect them to grow, can they be added as list to variables.bzl and then read using append_to_filelist macro?
```suggestion
      "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.cpp"
      "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_utils.cpp"
```

**Review by malfet — APPROVED** [review]

File: caffe2/CMakeLists.txt (hunk 1 above)
Comment (beauby) [reply caffe2/CMakeLists.txt:306]: As `"${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.h"` was already there I just followed the usage, but I can remove both headers. We don't expect that list to grow though.

**Comment by beauby** [comment]
CI is passing in https://github.com/pytorch/pytorch/pull/44947 (same exact commit).

File: torch/csrc/jit/serialization/pickler.h [hunk 2]
```diff
@@ -7,6 +7,7 @@
 #include <ATen/core/ivalue.h>
 #include <ATen/core/jit_type.h>
 #include <c10/util/ArrayRef.h>
+#include <torch/csrc/WindowsTorchApiMacro.h>
```
Comment (mrshenli) [inline torch/csrc/jit/serialization/pickler.h:10]: why do we need a windows header here?

**Review by mrshenli — APPROVED** [review]
RPC changes LGTM!

**Comment by mrshenli** [comment]
Curious, how did you trigger all tests without a ci-all branch? Is adding ci/all tag sufficient?

BTW, what's the difference between this one and #44947. Are test failures on this one expected?

File: torch/csrc/jit/serialization/pickler.h (hunk 2 above)
Comment (beauby) [reply torch/csrc/jit/serialization/pickler.h:10]: This is for `TORCH_API`, which resolves to `CAFFE2_API`, which resolves to `C10_EXPORT`, which resolves to `__attribute__((__visibility__("default")))` on linux, effectively exporting the symbols. Without the `TORCH_API`, `Pickler` and `Unpickler` are not being exported in libtorch.

**Comment by beauby** [comment]
@mrshenli #44947 points to the same commit as this branch. The reason it exists is that we were under the impression that tag `ci/all` only worked on PR for branches of the main repo. The failures here are cancellations of duplicate jobs.

**Pull request closed without merging**